cache/cache_reservation_manager.cc

```diff
@@ -0,0 +1,146 @@
+  if (new_mem_used == cur_cache_allocated_size ||
+      (cur_cache_allocated_size > 0 &&
+       new_mem_used > cur_cache_allocated_size &&
+       (new_mem_used - cur_cache_allocated_size) < kSizeDummyEntry) ||
+      (new_mem_used < cur_cache_allocated_size &&
+       (cur_cache_allocated_size - new_mem_used) < kSizeDummyEntry)) {
+    return Status::OK();
+  } else if (new_mem_used > cur_cache_allocated_size) {
+    Status s = IncreaseCacheReservation<R>(new_mem_used);
+    return s;
+  } else {
+    // In delayed decrease mode, we don't decrease cache reservation
+    // untill the memory usage is less than 3/4 of what we reserve
+    // in the cache.
+    // We do this because
+    // (1) Dummy entry insertion is expensive in block cache
+    // (2) Delayed releasing previously inserted dummy entries can save such
+    // expensive dummy entry insertion on memory increase in the near future,
+    // which is likely to happen when the memory usage is greater than or equal
+    // to 3/4 of what we reserve
+    if (delayed_decrease_ && new_mem_used >= cur_cache_allocated_size / 4 * 3) {
+      return Status::OK();
+    } else {
+      Status s = DecreaseCacheReservation(new_mem_used);
+      return s;
+    }
+  }
+}
+
+// Explicitly instantiate templates for "CacheEntryRole" values we use.
+// This makes it possible to keep the template definitions in the .cc file.
+template Status CacheReservationManager::UpdateCacheReservation<
+    CacheEntryRole::kWriteBuffer>(std::size_t new_mem_used);
+// For cache reservation manager unit tests
+template Status CacheReservationManager::UpdateCacheReservation<
+    CacheEntryRole::kMisc>(std::size_t new_mem_used);
+
+template <CacheEntryRole R>
+Status CacheReservationManager::IncreaseCacheReservation(
+    std::size_t new_mem_used) {
+  Status return_status = Status::OK();
+  while (new_mem_used > cache_allocated_size_.load(std::memory_order_relaxed)) {
+    Cache::Handle* handle = nullptr;
+    Status s = cache_->Insert(GetNextCacheKey(), nullptr, kSizeDummyEntry,
+                              GetNoopDeleterForRole<R>(), &handle);
+
+    // We do not return immeditately when encountering dummy entry insertion
+    // error to keep cache_allocated_size_ easy to reasonate. In other words, we
+    // still account those dummy entries that fail insertion in
+    // cache_allocation_size_. As a consequence, there will exist null dummy
+    // handles in dummy_handles_ if insertion error occurs.
+    //
+    // Also, We do not overwrite any non-OK status from dummy entry insertion.
+    // This is to prevent possible subsequent successful dummy entry insertion
+    // (due to newly available cache space resulting from manipulation in other
+    // threads) from overriding the non-OK status
+    if (return_status == Status::OK()) {
+      return_status = s;
+    }
+
+    // We absorb the error here since s is conditionally checked
+    // when return_status == Status::OK()
+    s.PermitUncheckedError();
```

Comment:
[Note]
I feel that L98 - 101 describes a valid concern and I haven't thought of another way to handle that. One drawback to this approach of preventing status from being overwritten is the need of s.PermitUncheckedError();. Let me know if L98 - 101 is not a valid concern and I can return s directly without the need for s.PermitUncheckedError();.

Reply:
I think return it directly is good. The old best-effort behavior of adding nullptr handles for failed insertions seems not generally useful. It seems questionable whether it's even useful for `WriteBufferManager` today -- and definitely not useful once we make `WriteBufferManager` able to handle reservation failures.

It would be a behavior change for `WriteBufferManager` in today's form, though. Failure to insert enough dummy entries in one `UpdateCacheReservation()` would start causing extra insertions (or fewer releases) in the next `UpdateCacheReservation()`. I think that's fine since we want to continue targeting an accurate reservation even after some failure.

The downside of this change is an individual `UpdateCacheReservation()` might reserve less because we fail early rather than continue trying more cache keys. It could be mitigated by retrying a few times in `WriteBufferManager` when a failure is returned. But I doubt it's needed.